adcc/ExcitedStates.py

```diff
@@ -153,33 +178,45 @@ def __init__(self, data, method=None, property_method=None,
         property_method : str, optional
             Provide an explicit method for property calculations to
             override the automatic selection.
-        excitation_energy_corrections : dict, optional
-            Provide a dictionary of functions to compute corrections for
-            excitation energies, called for each excitation individually
         """
         super().__init__(data, method, property_method)
-
-        if hasattr(data, "excitation_energy_corrections"):
-            merged = data.excitation_energy_corrections.copy()
-            merged.update(excitation_energy_corrections)
-            excitation_energy_corrections = merged
-
-        if hasattr(self.reference_state, "excitation_energy_corrections"):
-            merged = self.reference_state.excitation_energy_corrections.copy()
-            merged.update(excitation_energy_corrections)
-            excitation_energy_corrections = merged
-
-        self.excitation_energy_corrections = excitation_energy_corrections
-        for key in self.excitation_energy_corrections:
-            corr_function = self.excitation_energy_corrections[key]
-            if not callable(corr_function):
-                raise TypeError("Elements in excitation_energy_corrections "
-                                "must be callable.")
-            # call the function for exc. energy correction
-            correction = np.array([corr_function(exci)
-                                   for exci in self.excitations])
-            setattr(self, key, correction)
-            self._excitation_energy += correction
+        self._excitation_energy_corrections = []
+        # copy energy corrections if possible
+        # and avoids re-computation of the corrections
+        if hasattr(data, "_excitation_energy_corrections"):
+            for eec in data._excitation_energy_corrections:
+                correction_energy = getattr(data, eec.name)
+                setattr(self, eec.name, correction_energy)
```

Comment:
I think the pythonic way is to provoke a key error and catch it to do the caching.

Reply:
I'm not sure where a key error could/should be provoked in these lines?

Reply:
Ah sorry I misread the code.